samza-core/src/main/scala/org/apache/samza/storage/NonTransactionalStateTaskStorageManager.scala

```diff
@@ -0,0 +1,144 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.storage
+
+import java.io._
+
+import com.google.common.annotations.VisibleForTesting
+import com.google.common.collect.ImmutableSet
+import org.apache.samza.container.TaskName
+import org.apache.samza.job.model.TaskMode
+import org.apache.samza.system._
+import org.apache.samza.util.Logging
+import org.apache.samza.util.ScalaJavaUtil.JavaOptionals
+import org.apache.samza.{Partition, SamzaException}
+
+import scala.collection.JavaConverters._
+
+/**
+ * Manage all the storage engines for a given task
+ */
+class NonTransactionalStateTaskStorageManager(
+  taskName: TaskName,
+  containerStorageManager: ContainerStorageManager,
+  storeChangelogs: Map[String, SystemStream] = Map(),
+  systemAdmins: SystemAdmins,
+  loggedStoreBaseDir: File = new File(System.getProperty("user.dir"), "state"),
+  partition: Partition) extends Logging with TaskStorageManager {
+
+  private val storageManagerUtil = new StorageManagerUtil
+  private val persistedStores = containerStorageManager.getAllStores(taskName).asScala
+    .filter { case (storeName, storageEngine) => storageEngine.getStoreProperties.isPersistedToDisk }
+
+  def getStore(storeName: String): Option[StorageEngine] =  JavaOptionals.toRichOptional(containerStorageManager.getStore(taskName, storeName)).toOption
+
+  def flush(): Map[SystemStreamPartition, Option[String]] = {
+    debug("Flushing stores.")
+    containerStorageManager.getAllStores(taskName).asScala.values.foreach(_.flush)
+    val newestChangelogSSPOffsets = getNewestChangelogSSPOffsets()
+    writeChangelogOffsetFiles(newestChangelogSSPOffsets)
+    newestChangelogSSPOffsets
```

Comment:
Now it seems the newest changelog ssp offsets also contain changelog offsets for non-persisted store.  Shouldn't non-persisted store, e.g. in-memory store not have any checkpointed offset, rather start from oldest?

Reply:
The checkpointed offset is is the offset to restore _upto_. Starting offset is determined based on the oldest offset in the topic and the persisted offset file in the latest valid store checkpoint (if any). For logged in-memory stores, since there are no store checkpoints, TransactionalStateTaskRestoreManager will restore from oldest to checkpointed offset (line 302 in current impl).

These offsets aren't used in existing (non-transactional) restore manager, so will have no impact. 

Does this make answer your question?